### PR TITLE
store: Fix ephemeral behavior with Consul

### LINKED ISSUE #830: Default settings for Node removal are not optimal
Default settings for TTLs and Node removal on storage backends are not optimal and two annoying things could happen:

- It takes time using Consul and Etcd to detect a node failure, if the value of the TTL is high, swarm will see those nodes as alive until those entries are expired.
- This other scenario can happen:
  - Swarm agent registers onto discovery service
  - Manager sees the new entry: it is added to the list of known agents
  - Swarm agent fails
  - Manager removes the agent after the TTL expires
  - Swarm agent is back to life at the same time or directly after
  - It may take the Manager `heartbeat time` to register that node again although it's alive and working

One solution could be to reduce the default `heartbeat` and `refresh` time to a low value but at the risk of an increased pressure on the metadata storage cluster, especially if we use it for both discovery and metadata.

### PR DESCRIPTION
**Pull request by abronan**
Fixes #830 

This was an error with Consul: When we join a cluster with an agent, then kill it and re-join, the key is still there with the old unexpired session. But as we create a new one for ephemeral nodes each time we relaunch `swarm join`, the TTL for the old session expires and the Manager is removing the entry from the cluster because of the behavior set to `Delete`.

You then need to wait `heartbeat` time for this agent to join back the cluster with the new, correct session.

The solution is pretty primitive as we delete the key first to remove any attached session for that key. It's a quick fix but we need to find a better way to handle this gracefully.

Signed-off-by: Alexandre Beslic <abronan@docker.com>

### CODE REVIEW

**Comment by vieux** [comment]
ping @aluzzardi

File: pkg/store/consul.go [hunk 1]
```diff
@@ -97,9 +97,14 @@ func (s *Consul) setEphemeralTTL(ttl time.Duration) error {
...
 // CreateEphemeralSession creates the a global session
 // once that is used to delete keys at node failure
-func (s *Consul) createEphemeralSession() error {
+func (s *Consul) createEphemeralSession(key string) error {
 	s.Lock()
 	defer s.Unlock()
+
+	// Delete the key first (erase any attached session for that key)
+	// FIXME recover gracefully with the existing global session
+	s.Delete(key)
```
Comment (aluzzardi) [inline pkg/store/consul.go]: uhm, this will delete **every time**, not just the first time we create the ephemeral session - not sure if we could do it differently

Comment (abronan) [inline pkg/store/consul.go]: It deletes it a single time although it does not work as expected: each subsequent key going through the Put will stay with the old session (it thus only work truly with a single key, which is bad). A better solution is to get back the existing session but at the cost of the Manager not seeing the "change" of the Agent leaving and coming back to life.

I will keep this PR open and give it a shot.

**Comment by abronan** [comment]
@aluzzardi PTAL now the session is recovered instead of destroyed first. The agent just keeps refreshing the existing session `ttl` instead of a new one. Even though now the Manager does not see the failure (but it will mark the node as unhealthy anyway so it's not a problem).

**Comment by abronan** [comment]
Ok not good to review, still edge cases. **Do not merge** until this is fixed and there is proper integration test in #887.

File: pkg/store/consul.go [hunk 2]
```diff
@@ -95,11 +95,13 @@ func (s *Consul) setEphemeralTTL(ttl time.Duration) error {
 	return nil
 }
 
-// CreateEphemeralSession creates the a global session
+// CreateEphemeralSession creates the global session
```
Comment (aluzzardi) [inline pkg/store/consul.go]: `createEphemeralSession creates [...]`

File: pkg/store/consul.go [hunk 3]
```diff
@@ -108,11 +110,23 @@ func (s *Consul) createEphemeralSession() error {
...
+			return "", err
 		}
-		s.ephemeralSession = session
+		return session, nil
 	}
-	return nil
+	return "", nil
+}
+
+// checkPreviousSession checks if the key already has a session attached
```
Comment (aluzzardi) [inline pkg/store/consul.go]: `checkActiveSession checks...`

File: pkg/store/consul.go [hunk 4]
```diff
@@ -145,17 +159,32 @@ func (s *Consul) Put(key string, value []byte, opts *WriteOptions) error {
...
 		if s.ephemeralSession == "" {
-			s.createEphemeralSession()
+			if s.ephemeralSession, err = s.createEphemeralSession(); err != nil {
+				return err
+			}
+		}
+
+		// If a previous session is still active for that key, use it
+		// else we use the global ephemeral session
+		if previous != "" && previous != s.ephemeralSession {
```
Comment (aluzzardi) [inline pkg/store/consul.go]: there's probably no need to check `previous != s.ephemeralSession`

File: pkg/store/consul.go (hunk 2 above)
Comment (vieux) [inline pkg/store/consul.go]: @aluzzardi I agree for consistency, but this method is unexported. So golint won't complain

**Comment by abronan** [comment]
@aluzzardi PTAL

**Comment by aluzzardi** [comment]
LGTM
